src/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogTraceListener.cs

```diff
@@ -0,0 +1,174 @@
+﻿using System.Globalization;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System.Diagnostics
+{
+    public sealed class EventLogTraceListener : TraceListener
+    {
+        private bool _nameSet;
+        
+        public EventLogTraceListener()
+        {
+        }
+        
+        public EventLogTraceListener(EventLog eventLog)
+            : base(eventLog != null ? eventLog.Source : string.Empty)
+        {
+            EventLog = eventLog;
+        }
+        
+        public EventLogTraceListener(string source)
+        {
+            EventLog = new EventLog
+            {
+                Source = source
+            };
+        }
+        
+        public EventLog EventLog
+        {
+            get;
+            set;
+        }
+        
+        public override string Name
+        {
+            get
+            {
+                if (_nameSet == false && EventLog != null)
+                {
+                    _nameSet = true;
+                    base.Name = EventLog.Source;
+                }
+
+                return base.Name;
+            }
+
+            set
+            {
+                _nameSet = true;
+                base.Name = value;
+            }
+        }
+
+        public override void Close() => EventLog?.Close();
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                Close();
+                if (!disposing)
+                {
+                    EventLog = null;
+                }
+            }
+            finally
+            {
+                base.Dispose(disposing);
+            }
+        }
+
+        public override void Write(string message) => EventLog?.WriteEntry(message);
+        
+        public override void WriteLine(string message) => Write(message);
+
+        [ComVisible(false)]
+        public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType severity, int id, string format, params object[] args)
+        {
+            if (Filter != null && !Filter.ShouldTrace(eventCache, source, severity, id, format, args, null, null))
+                return;
+
+            EventInstance data = CreateEventInstance(severity, id);
+
+            if (args == null || args.Length == 0)
+            {
+                EventLog.WriteEvent(data, format);
```

Comment:
Just realized my previous question about args.Length might be problematic if format is null or empty.  What's supposed to happen if format is null or empty and args is null or empty... in that case we'll be calling EventLog.WriteEvent(data, null/"")... is that ok?

Reply:
I think it would be ok because we would be hitting: https://github.com/dotnet/corefx/blob/e0ba7aa8026280ee3571179cc06431baf1dfaaac/src/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs#L1379 -- then it would just create an empty array of strings, which would the same behavior we had earlier. We would create the empty array here: 

https://github.com/dotnet/corefx/blob/e0ba7aa8026280ee3571179cc06431baf1dfaaac/src/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs#L1401

I'm writing tests for this class, will add one with such scenario -- will create a new PR for the tests though.